buffer.go

```diff
@@ -103,14 +106,101 @@ func (b *Buffer) Serialize(serializationType SerializationType) ([]byte, error)
 	case TILEDB_CAPNP:
 		// The entire byte array contains Cap'nP data. Don't bother it.
 	case TILEDB_JSON:
-		// The data is a null-terminated string. Strip off the terminator.
+		// The data might be a null-terminated string. Strip off the terminator.
 		bs = bytes.TrimSuffix(bs, []byte{0})
 	default:
 		return nil, fmt.Errorf("unsupported serialization type: %v", serializationType)
 	}
 	return bs, nil
 }
 
+// ReadAt writes the contents of a Buffer at a given offset to a slice.
+func (b *Buffer) ReadAt(p []byte, off int64) (int, error) {
+	if off < 0 {
+		return 0, fmt.Errorf("offset cannot be negative")
+	}
+
+	var cbuffer unsafe.Pointer
+	var csize C.uint64_t
+
+	ret := C.tiledb_buffer_get_data(b.context.tiledbContext, b.tiledbBuffer, &cbuffer, &csize)
+	if ret != C.TILEDB_OK {
+		return 0, fmt.Errorf("error getting tiledb buffer data: %w", b.context.LastError())
+	}
+
+	if uintptr(off) > uintptr(csize) {
```

Comment:
Maybe, but I'm thinking whether it would be better to `return 0, io.EOF` here. [Documentation](https://pkg.go.dev/io#ReaderAt) is not clear on the behavior of `off` being out of bounds and I found [this SO question](https://stackoverflow.com/questions/76665282/behaviour-of-io-readat-at-the-end-of-input-source) which wasn't very clear either. BTW the OS file APIs (`pread` and `ReadFile`) do not fail; I ran the following C# program and it printed `0 0 0` on both Linux and Windows.

```csharp
// Creates an empty temporary file
string path = Path.GetTempFileName();

using var fh = File.OpenHandle(path, FileMode.Open, FileAccess.Read, options: FileOptions.DeleteOnClose);

byte[] buff = new byte[64];
Console.WriteLine(RandomAccess.Read(fh, buff, 0));
Console.WriteLine(RandomAccess.Read(fh, buff, 1));
Console.WriteLine(RandomAccess.Read(fh, buff, 1L << 32));
```

---

I later asked Gemini "In Go, what should the behavior of ReadAt be if the offset is out of bounds?" and it said:

> In Go, the `ReadAt` function should return an error if the offset is out of bounds. Specifically:
>
> 1. Offset is negative: It should return an error like `errors.New("offset cannot be negative")`.
>
> 2. Offset is beyond the buffer's size: It should return an error. A clear error message, like `fmt.Errorf("offset cannot be greater than buffer size")`, is helpful. If the offset equals the buffer's size, meaning there's nothing to read, it's acceptable (though not strictly required by the io.ReaderAt interface) to return 0, io.EOF to signal the end of the data.
>
> It's important _not_ to panic in these out-of-bounds situations. Return an error so the calling code can handle the situation gracefully.

But I don't know if it is biased on the code I already wrote. 😅

Reply:
`os.File.ReadAt()` returns `0, io.EOF` in case of offset overflows and fails in case of negative offsets, as evidenced by the following test:

```go
func TestReadAtFile(t *testing.T) {
	f, err := os.CreateTemp("/tmp", "*")
	require.NoError(t, err)
	defer os.Remove(f.Name())

	buff := make([]byte, 1024)
	n, err := f.ReadAt(buff, 0)
	require.Equal(t, io.EOF, err)
	n, err = f.ReadAt(buff, 1<<32)
	require.Equal(t, io.EOF, err)
	n, err = f.ReadAt(buff, -1)
	require.NotEqual(t, io.EOF, err)

	require.Equal(t, 0, n)
}
```

I updated `Buffer.ReadAt()` to have a similar behavior.